llvm-hs-pure/src/LLVM/IRBuilder/Monad.hs

```diff
@@ -154,9 +154,12 @@ freshName :: MonadIRBuilder m => ShortByteString -> m Name
 freshName suggestion = do
   usedNames <- liftIRState $ gets builderUsedNames
   let
-    candidates = suggestion : [suggestion <> fromString (show n) | n <- [(1 :: Int)..]]
-    (unusedName:_) = filter (not . (`HS.member` usedNames)) candidates
-  liftIRState $ modify $ \s -> s { builderUsedNames = HS.insert unusedName $ builderUsedNames s }
+    nameCount = fromMaybe 0 $ M.lookup suggestion usedNames
+    unusedName = if nameCount == 0
```

Comment:
I believe this does not guarantee that the name is unused:

1. Consider the case where we first lookup `a` twice so we have `a` and `a1`. Now the user uses `a1` as the suggested name, the `nameCount` for `a1` will be `0` so we get a collision.

2. A second somewhat orthogonal issue happens if you use `a1` twice so you end up with `a11` and `a` 12 times so you end up with `a11` again.

I think you can solve both of those issues by 1. appending something even if the name is used the first time and 2. adding a separator between the number and the name so you avoid interpreting digits at the end of the suggested name as part of the number suffix.

So maybe something like 
```
unusedName = suggestion <> "_" <> fromString (show nameCount)

Does that sound reasonable?

Reply:
Right, silly me.. I had the existing code too much in my mind, where the suggestion is prepended as is.
I will update it like you suggested, should be unique because of the extra "_".